src/Command/Geo/ZoneMatcherCommand.php

```diff
@@ -0,0 +1,185 @@
+use App\Entity\ZoneableEntity;
+use App\Geo\ZoneMatcher;
+use Doctrine\Bundle\DoctrineBundle\Repository\ServiceEntityRepository;
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Common\Collections\Collection;
+use Doctrine\ORM\EntityManagerInterface;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Output\OutputInterface;
+use Symfony\Component\Console\Style\SymfonyStyle;
+
+class ZoneMatcherCommand extends Command
+{
+    private const BULK = 500;
+
+    protected static $defaultName = 'app:geo:zone-matcher';
+
+    /**
+     * @var EntityManagerInterface
+     */
+    private $em;
+
+    /**
+     * @var ZoneMatcher
+     */
+    private $zoneMatcher;
+
+    /**
+     * @var SymfonyStyle
+     */
+    private $io;
+
+    /**
+     * @var Collection
+     */
+    private $entities;
+
+    public function __construct(EntityManagerInterface $em, ZoneMatcher $zoneMatcher)
+    {
+        $this->em = $em;
+        $this->zoneMatcher = $zoneMatcher;
+
+        parent::__construct();
+    }
+
+    protected function configure(): void
+    {
+        $this
+            ->setDescription('Assigns zones to entities')
+            ->addArgument(
+                'entity',
+                InputArgument::REQUIRED,
+                sprintf('Entity fully qualified class name (must implements %s and %s).', PostAddressHolderInterface::class, ZoneableEntity::class)
+            )
+            ->addOption('id', null, InputOption::VALUE_REQUIRED, 'Entity id.')
+            ->addOption('dry-run', null, InputOption::VALUE_NONE, 'Execute the algorithm but will not persist in database.')
+        ;
+    }
+
+    protected function initialize(InputInterface $input, OutputInterface $output): void
+    {
+        $this->io = new SymfonyStyle($input, $output);
+        $this->entities = new ArrayCollection();
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $this->io->title('Start assigning zones');
+
+        $class = $input->getArgument('entity');
+        $this->validateEntityClass($class);
+
+        $id = $input->getOption('id');
+
+        $dryRun = $input->getOption('dry-run');
+
+        $this->io->progressStart($id ? 1 : $this->countItems($class));
+
+        for ($offset = 0; $items = $this->getItems($class, $offset, $id); $offset += \count($items)) {
+            foreach ($items as $entity) {
+                $postalAddress = $entity->getPostAddress();
+                if (!$postalAddress) {
+                    continue;
+                }
+
+                $zones = $this->zoneMatcher->match($postalAddress);
+                if (0 === \count($zones)) {
+                    continue;
+                }
+
+                $this->applyZones($entity, $zones);
+                $this->em->persist($entity);
```

Comment:
not need to persist the entity as it is not a new entity, but already tracked by doctrine